ivm/src/ext.rs

```diff
@@ -5,13 +5,225 @@
 
 use core::{
   fmt::{self, Debug},
-  mem::transmute,
+  marker::PhantomData,
   ops::{Add, Div, Mul, Sub},
 };
-use std::io::{self, Read, Write};
+use std::{
+  collections::HashMap,
+  io::{self, Read, Write},
+};
 
 use crate::port::Tag;
 
+macro_rules! define_ext_fns {
+  ($map:ident,$self:ident, $($name:expr => |$a:ident, $b: ident| $body:expr ),*) => {
+    $($map.insert($name.to_string(), $self.register_ext_fn(move |[$a, $b]| [$body]).unwrap());)*
+  };
+}
+#[derive(Default)]
+pub struct Extrinsics<'ivm> {
+  ext_fns: Vec<Box<dyn Fn(ExtVal<'ivm>, ExtVal<'ivm>) -> ExtVal<'ivm> + Sync + 'ivm>>,
+  // amount of registered light exttys
+  light_ext_ty: u16,
+
+  n32_ext_ty: Option<ExtTy<'ivm>>,
+  io_ext_ty: Option<ExtTy<'ivm>>,
+
+  phantom: PhantomData<fn(&'ivm ()) -> &'ivm ()>,
+}
+
+impl<'ivm> Extrinsics<'ivm> {
+  pub const MAX_EXT_FN_KIND_COUNT: usize = 0x8FFF;
+  pub const MAX_LIGHT_EXT_TY_COUNT: usize = 0x8FFF;
+  pub const MAX_RC_EX_TTY_COUNT: usize = 0x8FFF;
+
+  pub fn register_ext_fn(
+    &mut self,
+    f: impl Fn([ExtVal<'ivm>; 2]) -> [ExtVal<'ivm>; 1] + Sync + 'ivm,
+  ) -> Option<ExtFn<'ivm>> {
+    if self.ext_fns.len() >= Self::MAX_EXT_FN_KIND_COUNT {
+      None
+    } else {
+      let ext_fn = ExtFn(self.ext_fns.len() as u16, PhantomData);
+      self.ext_fns.push(Box::new(move |a, b| f([a, b])[0]));
+      Some(ext_fn)
+    }
+  }
+  pub fn register_light_ext_ty(&mut self) -> Option<ExtTy<'ivm>> {
+    if self.light_ext_ty as usize >= Self::MAX_LIGHT_EXT_TY_COUNT {
+      None
+    } else {
+      let ext_ty = ExtTy::from_id_and_rc(self.light_ext_ty, false);
+      self.light_ext_ty += 1;
+      Some(ext_ty)
+    }
+  }
+  /// Some extrinsics are built-in with IVM
+  pub fn register_builtin_extrinsics(
```

Comment:
I think it would be nice if this could be moved out of `ivm` and into `ivy`. Then the intermediate `HashMap` wouldn't be necessary (the code in `ivy` could call methods on the host) and it would be easier to supply alternate implementations of these default extrinsics.